### ISSUE TITLE
Combine genesis test with invariant check

### ISSUE THREAD

**Issue by gjermundgaraba** [EmpowerPlastic/empowerchain#382]

**Comment by ba1ciu** [comment]
@gjermundgaraba 

My idea of reusing stuff we already implemented for genesis export would be to refactor getting all store records to iterators that use callback/handler functions like [we have in accesscontrol](https://github.com/EmpowerPlastic/empowerchain/blob/91ae1e730175243e08a2db940efd565c84ef7ed7/chain/x/accesscontrol/keeper/keeper.go#L73). Then we can just pass a handler that creates a slice of all objects in case of export and a different handler that checks invariant for invariant.go.

**Comment by gjermundgaraba** [comment]
Sounds perfect! LFG!

**Comment by ba1ciu** [comment]
@gjermundgaraba @maheshKRex let's list below what we think should be a system invariant for plastic credit

**Comment by ba1ciu** [comment]
1. Sum of all credit balances should equal a total number of credits in the credit collection.
2. ID Counters should point to next unexisting ID.

**Comment by gjermundgaraba** [comment]
I think we can break down the balance checks more too to also make sure the total number of credits within a denom is correct (balances.sum == collections.sum). Then a final total check.

Can't really think of anything else that would should be chain-halting. I assume if any links are broken for instance, the worst that will happen is that entities can't function (e.g. a project with a broken applicant_id link would just not be possible to use, right?), not that they would open any potential security vulns.

**Comment by ba1ciu** [comment]
Exactly. It will break using state that is corrupted in any way and shouldn't have any other consequences

**Comment by gjermundgaraba** [comment]
@ba1ciu Where are we on this one?

**Comment by ba1ciu** [comment]
So as far as I remember, when we discussed the matter with Marko, we agreed that only chain halting events business-wise should be implemented as invariants. I don't think there are any others than credit balance corruption?

**Comment by gjermundgaraba** [comment]
Perhaps also the next counters should not point to an existing entity? If it does I also think something is very bad and things can get corrupted?

**Comment by ba1ciu** [comment]
I thought about that, but wasn't sure if we actually need that as counters are so simple that it's going to be really hard to break them? Anyway, this invariant check won't be expensive so we can as well add it.

**Comment by gjermundgaraba** [comment]
True, but better to be on the safe side. This way, if we do mess them up, they would more easily be caught in simulation tests too.

**Comment by gjermundgaraba** [comment]
So let's do that as well in this issue, and then we can close it

**Comment by ba1ciu** [comment]
Closing the issue as per https://github.com/EmpowerPlastic/empowerchain/pull/530
